packages/SwingSet/src/kernel/liveSlots.js

```diff
@@ -453,50 +503,38 @@ function build(syscall, forVatID, cacheSize, vatPowers, vatParameters) {
     slotToVal.delete(promiseID);
   }
 
-  const ENABLE_PROMISE_ANALYSIS = true;
-
-  function retirePromiseIDIfEasy(promiseID, data) {
-    if (ENABLE_PROMISE_ANALYSIS) {
-      for (const slot of data.slots) {
-        const { type } = parseVatSlot(slot);
-        if (type === 'promise') {
-          lsdebug(
-            `Unable to retire ${promiseID} because slot ${slot} is a promise`,
-          );
-          return;
-        }
-      }
-    }
-    retirePromiseID(promiseID);
-  }
-
-  function thenResolve(promiseID) {
+  function thenHandler(p, promiseID, rejected) {
     insistVatType('promise', promiseID);
-    return res => {
-      harden(res);
-      lsdebug(`ls.thenResolve fired`, res);
-      const ser = m.serialize(res);
-      syscall.resolve([[promiseID, false, ser]]);
+    return value => {
+      knownResolutions.set(p, [rejected, value]);
+      harden(value);
```

Comment:
minor aesthetic thing, we should harden the value first, before placing it in knownResolutions.. doesn't affect the logic at all but makes our intentions clear

Reply:
Done.